src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -1060,7 +1066,12 @@ private void UpdateCompletedStreams()
                     }
 
                     _streams.Remove(stream.StreamId);
-                    ReturnStream(stream);
+
+                    // We don't want to return a stream for reuse that could still be draining
+                    if (stream.DrainExpirationTicks >= now)
+                    {
+                        ReturnStream(stream);
```

Comment:
If `stream.DrainExpirationTicks >= now` that means `stream.EndStreamReceived || stream.RstStreamReceived` in which case it should be perfectly fine to return the stream to the pool.

Reply:
Do you want that added as a comment?